operator/iamkeycontroller/create.go

```diff
@@ -54,45 +61,157 @@ func (p *IAMKeyPipeline) Create(ctx context.Context, mg resource.Managed) (manag
 
 // createIAMKey creates a new IAMKey in the project associated with the API Key and Secret.
 func (p *IAMKeyPipeline) createIAMKey(ctx *pipelineContext) error {
+	iamRoleID := ""
+	iamRoleNameFound := false
+
+	iamKeyFound := false
+
 	iamKey := ctx.iamKey
 	log := controllerruntime.LoggerFrom(ctx)
+	log.Info("starting creation")
+
+	// send request
+	resp, err := ExecuteRequest(ctx, "GET", ctx.iamKey.Spec.ForProvider.Zone, "/v2/iam-role", p.apiKey, p.apiSecret, nil)
+	if err != nil {
+		return err
+	}
+
+	iamRoles := IamRolesList{}
+	// unmarshall response to iamRoles
+
+	err = json.NewDecoder(resp.Body).Decode(&iamRoles)
+	if err != nil {
+		log.Error(err, "Cannot unmarshall response to iamRoles")
+		return err
+	}
+
+	for _, val := range iamRoles.IamRoles {
+		if *val.Name == iamKey.Spec.ForProvider.KeyName {
+			iamRoleNameFound = true
+			iamRoleID = *val.Id
+		}
+	}
 
-	var keyResources []exoscalesdk.IAMAccessKeyResource
-	for _, bucket := range iamKey.Spec.ForProvider.Services.SOS.Buckets {
-		keyResource := exoscalesdk.IAMAccessKeyResource{
-			Domain:       SOSResourceDomain,
-			ResourceName: bucket,
-			ResourceType: BucketResourceType,
+	policyDeny := exooapi.IamServicePolicyRuleActionDeny
+	policyRules := exooapi.IamServicePolicyTypeRules
+
+	// check if iam role exists if not create it
+	if !iamRoleNameFound {
+		log.Info("IAM Role doesnt exists, creating", "keyName", ctx.iamKey.Spec.ForProvider.KeyName)
+		autogeneratedAppcatRole := &exooapi.IamRole{
+			Name:        &iamKey.Spec.ForProvider.KeyName,
+			Description: pointer.String("IAM Role for SOS+IAM creation, it was autogenerated by provider-exoscale"),
+			Permissions: &[]exooapi.IamRolePermissions{
+				exooapi.IamRolePermissionsBypassGovernanceRetention,
+			},
+			Editable: pointer.Bool(true),
+			Policy: &exooapi.IamPolicy{
+				DefaultServiceStrategy: exooapi.IamPolicyDefaultServiceStrategyDeny,
+				Services: exooapi.IamPolicy_Services{
+					AdditionalProperties: map[string]exooapi.IamServicePolicy{
+						"sos": {
+							Type: &policyRules,
+							Rules: &[]exooapi.IamServicePolicyRule{
+								{
+									Action: &policyDeny,
+									// bucket name in this rule must be in single quotes, otherwise it doesn't work and user has access to everything
+									Expression: pointer.String("resources.bucket != " + "'" + strings.Split(iamKey.Spec.ForProvider.KeyName, ".")[1] + "'"),
+								},
+								{
+									Action:     &policyDeny,
```

Comment:
I read the support ticket you opened and it mentioned that it should be `policy allow = true`. I guess this should work as well?

Reply:
fixed